alphafold3_pytorch/alphafold3.py

```diff
@@ -2025,8 +2031,14 @@ def forward(
 
         rot_matrix[det_mask] = einsum(U[det_mask], V_fixed[det_mask], 'b i j, b j k -> b i k')
 
-        # Apply the rotation and translation
-        aligned_coords = einsum(pred_coords_centered, rot_matrix, 'b n i, b i j -> b n j') + true_centroid
+        # (A seeming typo in line 11 of Algorithm 28?
+        #  The bias should be \vec{\mu}^{GT} in the notation there,
+        #  I think.)
+
+        # Left multiplication of the ground truth by the rotation
+        # and then translation by the centroid of the prediction
+        # to "align" ground truth with prediction
+        aligned_coords = einsum(rot_matrix, true_coords_centered, 'b i j, b n j -> b n i') + pred_centroid
```

Comment:
This seems incorrect, in that the "lefthand" side of `rot_matrix` should represent the "coordinates axis" (when deriving `cov_matrix` by multiplying your target (to) coordinates by your source (from) coordinates). In the original version of this line, `pred_coords_centered` is correctly right-multiplied by the "lefthand" side (i.e., axis `i`) of `rot_matrix` which represents multiplication of the source coordinates by the "coordinates axis" of `rot_matrix`.

For example, if you instead compute `cov_matrix` by multiplying your source (from) coordinates by your target (to) coordinates (as in e.g., `H = torch.matmul(p.transpose(1, 2), q)` from https://hunterheidenreich.com/posts/kabsch_algorithm/), then to apply your `rot_matrix` you instead need to multiply your source coordinates by the transpose of `rot_matrix` (i.e., `torch.matmul(p, R.transpose(1, 2))` from the above link), which equivalently represents multiplying your source coordinates by the (now) "lefthand" coordinates axis of `rot_matrix`.

Reply:
I'll write up a unit test now to confirm all this.

Reply:
I've added a fix to the original `WeightedRigidAlign` module along with an updated unit test: https://github.com/lucidrains/alphafold3-pytorch/pull/22